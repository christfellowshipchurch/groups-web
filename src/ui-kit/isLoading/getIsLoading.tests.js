import React from 'react';
import renderer from 'react-test-renderer';

// import Providers from '../../Providers';

import { getIsLoading, IsLoadingContext } from '.';

describe('the getIsLoading HOC', () => {
  it('should return false from context', () => {
    const Context = getIsLoading(({ isLoading }) => isLoading.toString());

    const tree = renderer.create(
      <IsLoadingContext.Provider value={false}>
        <Context />
      </IsLoadingContext.Provider>
    );

    expect(tree).toMatchSnapshot();
  });
  it('should return true from context', () => {
    const Context = getIsLoading(({ isLoading }) => isLoading.toString());

    const tree = renderer.create(
      <IsLoadingContext.Provider value>
        <Context />
      </IsLoadingContext.Provider>
    );

    expect(tree).toMatchSnapshot();
  });
  it('should return the isLoading value (true) from the component', () => {
    const Context = getIsLoading(({ isLoading }) => isLoading.toString());

    const tree = renderer.create(
      <IsLoadingContext.Provider>
        <Context isLoading />
      </IsLoadingContext.Provider>
    );

    expect(tree).toMatchSnapshot();
  });
});
