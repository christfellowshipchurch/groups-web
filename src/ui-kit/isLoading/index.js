import React, { useContext } from 'react';

// Create our IsLoadingContext
const IsLoadingContext = React.createContext(false);

/* Safely checks for isLoading to make it easy to build components that respond to either props or
 * context for rendering placeholder/loading state. */
const getIsLoading = (Component) => (props) => {
  // does the component have an isLoading property?
  if (Object.hasOwnProperty.call(props, 'isLoading')) {
    return <Component {...props} />;
  }

  // No? That's fine we'll pull it from context
  const isLoading = useContext(IsLoadingContext);

  return <Component isLoading={isLoading} {...props} />;
};

// Allows a component to accept an `isLoading` prop to pass down into context for it's children.
const withIsLoading = (Component) => ({ isLoading, ...props }) => (
  <IsLoadingContext.Provider value={isLoading}>
    <Component {...props} />
  </IsLoadingContext.Provider>
);

export { getIsLoading, IsLoadingContext, withIsLoading };
