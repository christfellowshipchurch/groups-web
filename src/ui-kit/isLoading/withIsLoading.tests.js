import React from 'react';
import renderer from 'react-test-renderer';

import { getIsLoading, withIsLoading } from '.';

describe('the withIsLoading HOC', () => {
  it('should render', () => {
    const ParentComponentPassingConext = withIsLoading(
      ({ children }) => children
    );

    const tree = renderer.create(
      <ParentComponentPassingConext>
        <h1>Boom</h1>
      </ParentComponentPassingConext>
    );

    expect(tree).toMatchSnapshot();
  });
  it('should pass the isLoading prop (true) to context for children to consume', () => {
    const ChildComponentGettingContext = getIsLoading(({ isLoading }) =>
      isLoading.toString()
    );

    const ParentComponentPassingConext = withIsLoading(() => (
      <ChildComponentGettingContext />
    ));

    const tree = renderer.create(<ParentComponentPassingConext isLoading />);

    expect(tree).toMatchSnapshot();
  });
});
