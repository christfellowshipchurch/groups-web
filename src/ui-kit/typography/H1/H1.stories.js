import React from 'react';
import { storiesOf } from '@storybook/react';

import H1 from '.';

storiesOf('ui-kit/typography/H1', module)
  .add('default', () => (
    <H1>
      {
        '"Do all the good you can. By all the means you can. In all the ways you can. In all the places you can. At all the times you can. To all the people you can. As long as ever you can." ― John Wesley'
      }
    </H1>
  ))
  .add('padded', () => <H1 padded>Padded Heading Big and Long</H1>)
  .add('isLoading', () => (
    <H1 isLoading showLoadingAnimation>
      {
        '"Do all the good you can. By all the means you can. In all the ways you can. In all the places you can. At all the times you can. To all the people you can. As long as ever you can." ― John Wesley'
      }
    </H1>
  ))
  .add('isLoading + padded', () => (
    <H1 isLoading padded showLoadingAnimation>
      {
        '"Do all the good you can. By all the means you can. In all the ways you can. In all the places you can. At all the times you can. To all the people you can. As long as ever you can." ― John Wesley'
      }
    </H1>
  ));
