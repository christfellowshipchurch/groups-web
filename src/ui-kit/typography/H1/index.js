import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';

import { withOverrides } from '../../theme';
import { getIsLoading } from '../../isLoading';
import { Line } from '../../Placeholder';

const verticalRhythm = css`
  ${({ withMargins, theme }) =>
    withMargins ? `${theme.helpers.verticalRhythm(2.015625)} 0` : 'initial'};
`;

const StyledH1 = styled.h1.attrs(withOverrides('uiKit.typography.H1.StyledH1'))`
  font-size: 2.6875rem;
  font-weight: 900;
  line-height: ${({ theme }) => theme.helpers.verticalRhythm(2.6875, 1.15)};
  margin: ${verticalRhythm};

  /* Styles passed via the style prop and/or overrides */
  ${({ $style }) => $style}
`;

const Placeholder = styled(Line).attrs(
  withOverrides('uiKit.typography.H1.Placeholder')
)`
  height: 2.6875rem; /* Should match .h1 font-size */
  margin: ${({ withMargins, theme }) =>
    // if withMargins...
    withMargins
      ? // return rhythm so placeholder matches text position to prevent jitter/jank
        verticalRhythm
      : // else default to...
        `calc(${theme.helpers.verticalRhythm(
          2.6875,
          1.15
        )} - 2.6875rem) 0`}; /* line-height - font-size = space between lines */

  /* Styles passed via overrides */
  ${({ $style }) => $style}
`;

// We pass `style` as a prop to `StyledH2` so that custom styles are handled by styled components and not react (inline style)
const H1 = getIsLoading(
  ({ children, isLoading, withMargins, style, ...props }) => (
    <Placeholder
      isLoading={isLoading}
      withMargins={withMargins}
      showLoadingAnimation
    >
      <StyledH1 withMargins={withMargins} $style={style} {...props}>
        {children}
      </StyledH1>
    </Placeholder>
  )
);

H1.propTypes = {
  children: PropTypes.node.isRequired,
  isLoading: PropTypes.bool,
  withMargins: PropTypes.bool,
  style: PropTypes.any, // eslint-disable-line
};

H1.displayName = 'ui-kit.typography.H1';

export default styled(H1).attrs(withOverrides('uiKit.typography.H1'))``;
