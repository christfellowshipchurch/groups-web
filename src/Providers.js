import React from 'react';
import { ThemeProvider } from 'styled-components';
import PropTypes from 'prop-types';

import { defaultTheme } from './ui-kit/theme';

const Providers = ({ children }) => (
  <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
);

Providers.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

export default Providers;
