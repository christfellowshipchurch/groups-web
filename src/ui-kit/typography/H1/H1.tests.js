import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import Providers from '../../../Providers';

import { IsLoadingContext } from '../../isLoading';

import H1 from '.';

describe('the H1 component', () => {
  it('should render', () => {
    const tree = renderer.create(
      <Providers>
        <H1>Default H1 text</H1>
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
  it('should accept and render passed in styles', () => {
    const salmonText = { color: 'salmon' };
    const tree = renderer.create(
      <Providers>
        <H1 style={salmonText}>Salmon text</H1>
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
  it('should render with vertical rhythm (padded)', () => {
    const tree = renderer.create(
      <Providers>
        <H1 padded>H1 text with vertical rhythm styled via margin.</H1>
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
  it('should render a loading state', () => {
    const tree = renderer.create(
      <Providers>
        <H1 isLoading>Default H1 text</H1>
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
  it('should render a loading state via context', () => {
    const tree = renderer.create(
      <Providers>
        <IsLoadingContext.Provider value>
          <H1>Default H1 text</H1>
        </IsLoadingContext.Provider>
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
  it('should render a loading state with vertical rhythm (padded)', () => {
    const tree = renderer.create(
      <Providers>
        <H1 isLoading padded>
          Default H1 text
        </H1>
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
});
