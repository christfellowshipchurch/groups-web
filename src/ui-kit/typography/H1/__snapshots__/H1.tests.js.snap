// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the H1 component should accept and render passed in styles 1`] = `
.c0 {
  font-size: 2.6875rem;
  font-weight: 900;
  line-height: 3.0906249999999997rem;
  margin: initial;
  color: salmon;
}

<h1
  className="c0 sc-1dj0tjq-2 kEIKSh"
>
  Salmon text
</h1>
`;

exports[`the H1 component should render 1`] = `
.c0 {
  font-size: 2.6875rem;
  font-weight: 900;
  line-height: 3.0906249999999997rem;
  margin: initial;
}

<h1
  className="c0 sc-1dj0tjq-2 kEIKSh"
>
  Default H1 text
</h1>
`;

exports[`the H1 component should render a loading state 1`] = `
.c0 {
  background: #A5A5A5;
  border-radius: 16px;
  display: block;
  height: 1rem;
}

.c1 {
  height: 2.6875rem;
  margin: calc(3.0906249999999997rem - 2.6875rem) 0;
}

<mark
  className="c0 show-loading-animation ra2rjz-1 c1 custom-placeholder"
/>
`;

exports[`the H1 component should render a loading state via context 1`] = `
.c0 {
  background: #A5A5A5;
  border-radius: 16px;
  display: block;
  height: 1rem;
}

.c1 {
  height: 2.6875rem;
  margin: calc(3.0906249999999997rem - 2.6875rem) 0;
}

<mark
  className="c0 show-loading-animation ra2rjz-1 c1 custom-placeholder"
/>
`;

exports[`the H1 component should render a loading state with vertical rhythm (withMargins) 1`] = `
.c0 {
  background: #A5A5A5;
  border-radius: 16px;
  display: block;
  height: 1rem;
}

.c1 {
  height: 2.6875rem;
  margin: 3.0234375rem 0;
}

<mark
  className="c0 show-loading-animation ra2rjz-1 c1 custom-placeholder"
/>
`;

exports[`the H1 component should render with vertical rhythm (withMargins) 1`] = `
.c0 {
  font-size: 2.6875rem;
  font-weight: 900;
  line-height: 3.0906249999999997rem;
  margin: 3.0234375rem 0;
}

<h1
  className="c0 sc-1dj0tjq-2 kEIKSh"
>
  H1 text with vertical rhythm styled via margin.
</h1>
`;
